# Intro to Data Visualization with Python

# Contents
1. Import libraries
2. Import datasets
3. Check imported data
4. Data wrangling
5. Consistency checks
6. Merge dataframes
7. Export dataframe

# 1. Import libraries

In [5]:
#Import libraries
import pandas as pd
import numpy as np
import os

# 2. Import datasets

In [7]:
#Define the project path as a string and assign to a variable called path
path = r'/Users/davidgriesel/Documents/0 - CareerFoundry/02 - Data Analytics Immersion/04 - Python Fundamentals for Data Analysts/Instacart Basket Analysis - IC 202409'

## 2.1. Merged data

In [9]:
#Import orders_products_merged_derived dataset from exercise 4.7
df_orders_products_merged = pd.read_pickle(os.path.join(path, '02 - Data', 'Prepared Data', '08_orders_products_grouped.pkl'))

## 2.2. Customers

In [11]:
#Import the customer dataset
df_customers = pd.read_csv(os.path.join(path, '02 - Data', 'Original Data', 'customers.csv'), index_col = False)

# 3. Check imported data

## 3.1. Merged data

In [14]:
#Get the dimensions of the dataframe
df_orders_products_merged.shape

(32404859, 24)

In [15]:
#Display the first 5 rows of the dataframe
df_orders_products_merged.head()

order_id  user_id  user_order_number  orders_day_of_week  \
0   2539329        1                  1                   2   
1   2539329        1                  1                   2   
2   2539329        1                  1                   2   
3   2539329        1                  1                   2   
4   2539329        1                  1                   2   

   order_hour_of_day  days_since_prior_order  first_order  product_id  \
0                  8                     NaN         True         196   
1                  8                     NaN         True       14084   
2                  8                     NaN         True       12427   
3                  8                     NaN         True       26088   
4                  8                     NaN         True       26405   

   add_to_cart_order  reordered  ...        price_label     busiest_day  \
0                  1          0  ...  Mid-range product  Regularly busy   
1                  2          0  ...  Mid-range product  Regularly busy   
2                  3          0  ...  Low-range product  Regularly busy   
3                  4          0  ...  Low-range product  Regularly busy   
4                  5          0  ...  Low-range product  Regularly busy   

     busiest_days   busiest_hours max_order  loyalty_flag avg_price  \
0  Regularly busy  Average orders        10  New customer  6.367797   
1  Regularly busy  Average orders        10  New customer  6.367797   
2  Regularly busy  Average orders        10  New customer  6.367797   
3  Regularly busy  Average orders        10  New customer  6.367797   
4  Regularly busy  Average orders        10  New customer  6.367797   

  spending_flag  median_days         frequency_flag  
0   Low spender         20.5  Non-frequent customer  
1   Low spender         20.5  Non-frequent customer  
2   Low spender         20.5  Non-frequent customer  
3   Low spender         20.5  Non-frequent customer  
4   Low spender         20.5  Non-frequent customer  

[5 rows x 24 columns]

## 3.2. Customers

In [17]:
#Dimensions
df_customers.shape

(206209, 10)

In [18]:
#Display the first 5 rows of the dataframe
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

# 4. Data wrangling

## 4.1. Rename columns

In [21]:
#Display the first 5 rows of the dataframe
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [22]:
#Renaming columns that are inconsistent
df_customers.rename(columns = {
    'First Name': 'first_name', 
    'Surnam': 'surname',
    'Gender': 'gender',
    'STATE': 'state',
    'Age': 'age',
    'n_dependants': 'number_of_dependents',
    'fam_status': 'marital_status'
}, inplace = True)

In [23]:
#View results
df_customers.head()

user_id first_name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependents marital_status  income  
0                     3        married  165665  
1                     0         single   59285  
2                     2        married   99568  
3                     0         single   42049  
4                     1        married   40374

## 4.2. Drop columns

In [25]:
#View results
df_customers.isnull().sum()

user_id                     0
first_name              11259
surname                     0
gender                      0
state                       0
age                         0
date_joined                 0
number_of_dependents        0
marital_status              0
income                      0
dtype: int64

##### The first_name column contains 11259 null values. Non vital information and not enough records to validate deletion. No other columnsidentified for deletion at this stage.

## 4.3. Change data types

In [28]:
#View descriptive stats
df_customers.head()

user_id first_name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependents marital_status  income  
0                     3        married  165665  
1                     0         single   59285  
2                     2        married   99568  
3                     0         single   42049  
4                     1        married   40374

In [29]:
#View data types
df_customers.dtypes

user_id                  int64
first_name              object
surname                 object
gender                  object
state                   object
age                      int64
date_joined             object
number_of_dependents     int64
marital_status          object
income                   int64
dtype: object

##### No columns require changing data types

# 5. Consistency checks

## 5.1. Accuracy

### 5.1.1. View descriptive statistics

In [34]:
df_customers.describe()

user_id            age  number_of_dependents         income
count  206209.000000  206209.000000         206209.000000  206209.000000
mean   103105.000000      49.501646              1.499823   94632.852548
std     59527.555167      18.480962              1.118433   42473.786988
min         1.000000      18.000000              0.000000   25903.000000
25%     51553.000000      33.000000              0.000000   59874.000000
50%    103105.000000      49.000000              1.000000   93547.000000
75%    154657.000000      66.000000              3.000000  124244.000000
max    206209.000000      81.000000              3.000000  593901.000000

##### age: Customers range between 18 and 81 years. With mean and median of 49. 
##### number_of_dependants: Range between 0 and 3 with most customers having 1 dependant. 
##### income: Customer income varies between 25903 and 593901 with average customer earning in the low 90's. Wide range in terms of income. 
##### Data set has 206209 total records which agree with number of first orders per order dataset. 

## 5.2. Missing values

### 5.2.1. Find missing values

In [38]:
#Find missing values
df_customers.isnull().sum()

user_id                     0
first_name              11259
surname                     0
gender                      0
state                       0
age                         0
date_joined                 0
number_of_dependents        0
marital_status              0
income                      0
dtype: int64

In [39]:
df_customers_nan = df_customers[df_customers['first_name']. isnull() == True]

In [40]:
#View results
df_customers_nan

user_id first_name   surname  gender         state  age date_joined  \
53        76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73        13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82        89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99        96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105       29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...         ...        ...       ...     ...           ...  ...         ...   
206038   121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044   200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090   167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162   187532        NaN     Floyd  Female    California   39    4/1/2020   
206171   116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        number_of_dependents marital_status  income  
53                         2        married   41709  
73                         0         single   82518  
82                         3        married  117099  
99                         1        married  155673  
105                        3        married  151819  
...                      ...            ...     ...  
206038                     3        married   87783  
206044                     2        married  108488  
206090                     1        married   45275  
206162                     0         single   56325  
206171                     2        married   59222  

[11259 rows x 10 columns]

### 5.2.2 Address missing values

##### The first_name column contains 11259 null values. Non vital information and not enough records to validate deletion. No need to add a flag column and cannot substitue with measures of central tendancy as it is categorical.
##### No other columns identified for deletion at this stage.

## 5.3. Duplicates

### 5.3.1. Find duplicates

In [45]:
#Find duplicates and create subset with the results
df_customers_duplicates = df_customers[df_customers.duplicated()]

In [46]:
#View results
df_customers_duplicates

Empty DataFrame
Columns: [user_id, first_name, surname, gender, state, age, date_joined, number_of_dependents, marital_status, income]
Index: []

In [47]:
#Get the dimensions of the subset
df_customers_duplicates.shape

(0, 10)

### 5.3.2. Address duplicates

##### No duplicates identified

## 5.4. Mixed-type data

### 5.4.1. Find mixed-type data

In [52]:
for col in df_customers.columns.tolist():
    weird = (df_customers[[col]].map(type) != df_customers[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_customers[weird]) > 0:
        print (col)

first_name


In [53]:
#Return the data types of each column in the dataframe
df_customers.dtypes

user_id                  int64
first_name              object
surname                 object
gender                  object
state                   object
age                      int64
date_joined             object
number_of_dependents     int64
marital_status          object
income                   int64
dtype: object

In [54]:
#Change data type of mix column to string and update the dataframe with results
df_customers['first_name'] = df_customers['first_name'].astype('str')

### 5.4.2. Address mixed-type data

##### The first_name column was returned due to null values. No need to change data type

# 6. Merge dataframes

In [58]:
#Display the first 5 rows of the dataframe
df_customers.head()

user_id first_name   surname  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   number_of_dependents marital_status  income  
0                     3        married  165665  
1                     0         single   59285  
2                     2        married   99568  
3                     0         single   42049  
4                     1        married   40374

In [59]:
#Display the first 5 rows of the dataframe
df_orders_products_merged.head()

order_id  user_id  user_order_number  orders_day_of_week  \
0   2539329        1                  1                   2   
1   2539329        1                  1                   2   
2   2539329        1                  1                   2   
3   2539329        1                  1                   2   
4   2539329        1                  1                   2   

   order_hour_of_day  days_since_prior_order  first_order  product_id  \
0                  8                     NaN         True         196   
1                  8                     NaN         True       14084   
2                  8                     NaN         True       12427   
3                  8                     NaN         True       26088   
4                  8                     NaN         True       26405   

   add_to_cart_order  reordered  ...        price_label     busiest_day  \
0                  1          0  ...  Mid-range product  Regularly busy   
1                  2          0  ...  Mid-range product  Regularly busy   
2                  3          0  ...  Low-range product  Regularly busy   
3                  4          0  ...  Low-range product  Regularly busy   
4                  5          0  ...  Low-range product  Regularly busy   

     busiest_days   busiest_hours max_order  loyalty_flag avg_price  \
0  Regularly busy  Average orders        10  New customer  6.367797   
1  Regularly busy  Average orders        10  New customer  6.367797   
2  Regularly busy  Average orders        10  New customer  6.367797   
3  Regularly busy  Average orders        10  New customer  6.367797   
4  Regularly busy  Average orders        10  New customer  6.367797   

  spending_flag  median_days         frequency_flag  
0   Low spender         20.5  Non-frequent customer  
1   Low spender         20.5  Non-frequent customer  
2   Low spender         20.5  Non-frequent customer  
3   Low spender         20.5  Non-frequent customer  
4   Low spender         20.5  Non-frequent customer  

[5 rows x 24 columns]

In [60]:
#Merge dataframes using order_id and inner join while adding a merge flag
df_final_merged = df_orders_products_merged.merge(df_customers, on = 'user_id', indicator = True)

In [61]:
#Check merge rate
df_final_merged['_merge'].value_counts()

_merge
both          32404859
left_only            0
right_only           0
Name: count, dtype: int64

In [62]:
#Get the dimensions of the dataframe
df_final_merged.shape

(32404859, 34)

In [63]:
#Display the first 5 rows of the dataframe
df_final_merged.head()

order_id  user_id  user_order_number  orders_day_of_week  \
0   2539329        1                  1                   2   
1   2539329        1                  1                   2   
2   2539329        1                  1                   2   
3   2539329        1                  1                   2   
4   2539329        1                  1                   2   

   order_hour_of_day  days_since_prior_order  first_order  product_id  \
0                  8                     NaN         True         196   
1                  8                     NaN         True       14084   
2                  8                     NaN         True       12427   
3                  8                     NaN         True       26088   
4                  8                     NaN         True       26405   

   add_to_cart_order  reordered  ... first_name  surname  gender    state age  \
0                  1          0  ...      Linda   Nguyen  Female  Alabama  31   
1                  2          0  ...      Linda   Nguyen  Female  Alabama  31   
2                  3          0  ...      Linda   Nguyen  Female  Alabama  31   
3                  4          0  ...      Linda   Nguyen  Female  Alabama  31   
4                  5          0  ...      Linda   Nguyen  Female  Alabama  31   

  date_joined number_of_dependents marital_status  income _merge  
0   2/17/2019                    3        married   40423   both  
1   2/17/2019                    3        married   40423   both  
2   2/17/2019                    3        married   40423   both  
3   2/17/2019                    3        married   40423   both  
4   2/17/2019                    3        married   40423   both  

[5 rows x 34 columns]

# 7. Export dataframe

In [65]:
#Drop merge flag
df_final_merged = df_final_merged.drop(columns = ['_merge'])

In [66]:
#Check dimensions of the dataframe
df_final_merged.shape

(32404859, 33)

In [67]:
#Export dataframe
df_final_merged.to_pickle(os.path.join(path, '02 - Data','Prepared Data', '09_final_merged.pkl'))

# Task

## 1. Download dataset
Download the customer data set and add it to your “Original Data” folder.

##### Data downloaded

## 2. Create notebook
Create a new notebook in your “Scripts” folder for part 1 of this task.

##### Notebook created

## 3. Importing libraries
Import your analysis libraries, as well as your new customer data set as a dataframe.

##### Refer: 1. Import libraries
##### Refer: 2. Import datasets

## 4. Data wrangling
Wrangle the data so that it follows consistent logic; for example, rename columns with illogical names and drop columns that don’t add anything to your analysis.

##### Refer: 4. Data wrangling

## 5. Quality and consistency checks
Complete the fundamental data quality and consistency checks you’ve learned throughout this Achievement; for example, check for and address missing values and duplicates, and convert any mixed-type data.

##### Refer: 5. Consistency checks

## 6. Combine data
Combine your customer data with the rest of your prepared Instacart data. (Hint: Make sure the key columns are the same data type!)

##### Refer: 6. Merge dataframes

## 7. Check notebook
Ensure your notebook contains logical titles, section headings, and descriptive code comments.

##### Checked notebook, included section headings and code comments

## 8. Export dataframe
Export this new dataframe as a pickle file so you can continue to use it in the second part of this task.

##### Refer: 7. Export dataframe

## 9. Save & submit
Save your notebook so that you can send it to your tutor for review after completing part 2.

##### Notebook saved and submitted